.github/workflows/coverage_runner.yml

```diff
@@ -0,0 +1,109 @@
+name: Run tests, measure coverage and upload results to codecov
+on:
+  workflow_dispatch:
+    inputs:
+      pr_number:
+        description: Enter PR number coming from community to run coverage on it. Do not run for PR coming from hazelcast organization, they will be run automatically.
+        required: true
+  push:
+    branches:
+      - master
+      - '[45].*.z'
+  pull_request_target:
+    branches:
+      - master
+      - '[45].*.z'
+    types:
+      [opened]
+jobs:
+  check_for_membership:
+    runs-on: ubuntu-latest
+    name: Check membership of given user  
+    outputs:
+      check-result: ${{ steps.composite.outputs.check-result }}
+    steps: 
+      - name: Action for membership check
+        id: composite
+        uses: hazelcast/hazelcast-tpm/membership@main
+        with:
+          organization-name: 'hazelcast'
+          member-name: ${{ github.event.pull_request.head.repo.owner.login }}
+          token: ${{ secrets.PAT }}
+          
+  run-tests:
+    runs-on: ${{ matrix.os }}
+    needs: [check_for_membership]
+    if: github.event_name == 'push' || needs.check_for_membership.outputs.check-result == 'true' || github.event_name == 'workflow_dispatch'
+    name: Run tests with Python ${{ matrix.python-version }} on ${{ matrix.os }}
+    strategy:
+      matrix:
+        python-version: [ '3.6', '3.10' ]
+        os: [ ubuntu-latest, windows-latest ]
+      fail-fast: false
+    steps:
+      - name: Setup Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install JDK
+        uses: actions/setup-java@v2
+        with:
+          distribution: 'adopt'
+          java-version: '8'
+          
+      - name: Checkout code for push event
+        if: github.event_name == 'push'
+        uses: actions/checkout@v2
+
+      - name: Checkout code for PR
+        if: github.event_name == 'pull_request_target'
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+          
+      - name: Checkout repository to fetch PR coming from community
+        if: github.event_name == 'workflow_dispatch'
+        uses: actions/checkout@v2
```

Comment:
Why do we have three check out steps? The first and third can be merged. Can you explain why 2 is different?

Reply:
I felt to separate steps for every event trigger would be more understandable but yes we can merge it and I merged it at https://github.com/hazelcast/hazelcast-python-client/pull/539/commits/7cb06a7488d96c1f42d758b6ac12900be23a3051
Second one checking out the PR automatically.
First and third one checking out the base branch. If it is a `workflow_dispatch` so if it is triggered by us, in the next step, a new temp `pr_head` branch will be fetched.

Reply:
> First and third one checking out the base branch

Do not they check out the merge commit ? 🤔

Reply:
> Second one checking out the PR automatically.

Ok as I see in https://github.com/actions/checkout, the default value for `ref` is the sha of the event:

> \# The branch, tag or SHA to checkout. When checking out the repository that
> \# triggered a workflow, this defaults to the reference or SHA for that event.
> \# Otherwise, uses the default branch.
> ref: ''

And you can see that the GITHUB_SHA of [pull_request_target](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) event is "Last commit on the PR base branch", and the GITHUB_SHA of [pull_request](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request) event is 
"Last merge commit on the GITHUB_REF branch". 

So I think we should try to checkout the merge commit again for the pull_request_target event. I think right now we are checking the pull_request head commit which is not the same thing. I think we should use ${{ github.event.pull_request.merge_commit_sha }}.

Reply:
>Do not they check out the merge commit ?
Checking out the merge commit and base branch is the same thing I think

>So I think we should try to checkout the merge commit again for the pull_request_target event
Why? We are trying to run test on the PR, why should we checkout the merge commit?

BTW I already tested these checks out on github action machines and AFAIR it was working as expected. But I will double check it

Reply:
> Checking out the merge commit and base branch is the same thing I think

If checking out the base branch means checking out pull request's HEAD - I believe it does mean - the two are not the same thing. 

> Why? We are trying to run test on the PR, why should we checkout the merge commit?

Because that's the default behaviour of github. We may encounter with strange bugs. Github runs the tests on the merge commit because checkout action checks out to that commit. So if there are some changes in master that are not in the PR, those changes will be included in the test run as well. If you run it in the PR head, which is not the same thing as merge commit, we will have to merge master to pull request branch whenever master is updated. 

Workflows are running on merge commit because it proves the compatiblity when the changes of the PR is merged with the latest master. The merge commit is the new master commit after the merge which is reasonable to test.

Reply:
I got your point.
I tried to handle it with `${{ github.event.pull_request.merge_commit_sha }}` but it checks out the latest commit of the base branch not merge commit. It didn't work as expected. Maybe it is related with `pull_request_target`
But [https://stackoverflow.com/questions/63594658/git-refs-merge-vs-head-in-pull-request#:~:text=The%20refs%2Fpull%2F%3Cnumber,as%20a%20%22future%22%20commit.](this) is what we need I think. I already tested it and it seems it is the thing we need. Lets have a demo togather to be sure